### PR TITLE
Added dimension importance support to Decision Trees

### PR DESCRIPTION
**Pull request by TirelessClock**
Cherry-picked from PR #3736 

After training, xgboost can calculate feature importance to understand 
the contribution of each feature in the classification decision.
XGBoost provides two main types of feature importance scores:

- Weight (Frequency): This is the number of times a feature is used to 
  split a node across all trees in the model. It counts how often each 
  feature appears in all trees of the model.

- Gain (Cover): This measures the improvement in accuracy brought by a 
  feature to the model. For each feature, it sums the improvement in accuracy 
  (reduction in loss) brought by the feature when it is used in tree splits.

### CODE REVIEW

File: src/mlpack/methods/decision_tree/decision_tree.hpp [hunk 1]
```diff
@@ -496,6 +502,10 @@ class DecisionTree :
   std::vector<DecisionTree*> children;
   //! The dimension this node splits on.
   size_t splitDimension;
+  //! Stores the number of times a particular feature was used to split
+  map<size_t, size_t> featureFrequency;
+  //! Stores the net gain provided by a particular feature for splitting
+  map<size_t, double> featureCover;
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/decision_tree.hpp:508]: I don't have the time to review this PR deeply but I saw it touched this file, so I took a look: we can't add members to `DecisionTree`---these `map`s will be duplicated across every single node in the tree and will make the serialized model orders of magnitude larger (and predictions will be slower too).  An alternate approach will be needed.

Comment (TirelessClock) [reply src/mlpack/methods/decision_tree/decision_tree.hpp:508]: Oh right yes. I'll change it right away.
Thanks!

**Comment by rcurtin** [comment]
The actual implementation here is super simple (it's like two lines, you're just collecting simple information while building a tree), so I hate to keep bouncing this back about seemingly unrelated things, but I think there needs to be some deep discussion about design here.  Implementing XGBoost on top of the mlpack `DecisionTree` class is definitely the right idea, but we have to think carefully about how all of the pieces fit together.

Specifically we want to balance the following things:

 * So far as is possible, don't change the existing user-facing API for `DecisionTree`.  Adding to the API is okay, but we really need to manage the complexity as much as we can.
 * Changes should not affect the runtime of `DecisionTree` when they are specific to `XGBoost`.
 * If a new feature is being added, it should be considered how the feature could work with the existing `DecisionTree` class.
 * All documentation has to be updated accordingly.  A new feature that isn't documented effectively doesn't exist.

(For what it's worth, those four points are pretty much the thinking I am applying to all the XGBoost PRs.  These are hard things to get right, especially given how long it takes to get good context for mlpack and the components inside of it.  I'm doing my best to help with the design process.)

I would need to understand more about the feature here to give input on what the design should be.  Giving users the ability to compute feature importance for decision trees would be nice, but there are lots of measures of feature importance---the XGBoost strategy may not be the only one.  So if we do implement something, it should be general enough that we can expand it in the future.

The next question is: when is this measure used?  Is this only used after the XGBoost model is finished training?  If so, you can have a standalone `FeatureImportances()` method (or similar) that computes them on-the-fly, instead of computing them during training.

Can you elaborate on how this will be used and what the goals are?  You can see that there are a lot of questions to answer when adding a feature like this.

**Comment by TirelessClock** [comment]
Hello @rcurtin 

Sorry if some of these are a bit vague! I'll try to put down a bit of an introduction to my thought process so we can be on the same page. 

XGBoost is an extension to Gradient Boosting, which uses smaller, weak learner Decision Trees in ensemble to improve performance.

What's special about XGBoost is that it's not one specific algorithm, my understanding of it is that it takes salient feature of Gradient Feature and adds a whole bunch of additional features to it. I've implemented Gradient Boosting so far, and am now working towards implementing those appendage features for XGB. The features I'm aiming at right now are:

- Regularization (L1 and L2) 
- Tree Pruning 
- Support for Various Loss Functions 
- Feature Importance Calculation 
- Tree Method Options (exact, approx, hist) 
- Missing Value Handling 

As you can see, a lot of these need to be implemented at the Decision Tree level, and I can't figure out a upper layer implementation for, (for instance) Tree Pruning.

Now to address Feature Importance Calculation specifically, I know Feature Selection algos can be implemented generally, but what's specific to the XGBoost implementation is that it uses the gain calculated during tree building to rank the feature importance. 

I understand your concern about maintaining DecisionTree complexity, but if I tried to implement XGB specific trees separately, I'm worried about code redundancy, because much of the work is already done in the DecisionTree module. Though of course, it's possible if necessary.

Please read through this and share your thoughts.
Thanks!

**Comment by rcurtin** [comment]
So the question for each of those six features is, can you implement it without causing significant problems for the existing design of the `DecisionTree` class?  I think the answer is generally yes, but, it could be tricky to figure out how.  So, considering only the feature importance problem:

 * Adding a member function to `DecisionTree` is not *necessarily* a problem, however:
   - We can't add any member variables to `DecisionTree` because it is very important that structure remains small.  (Plus, the members would be duplicated at every level of the tree, which is not needed here.)
   - We want to make sure we have support for alternate measures of feature importance; so, one idea here is to use a template class to specify the feature importance type... `result = dt.FeatureImportances<XGBoostFeatureImportance>()`, and set the default to something reasonable.  Other measures could be implemented later.
   - We need to ensure that documentation is added to `decision_tree.md` accordingly.
   - This method will always have the drawback that the feature importances are not computed at training time; they would be computed as a separate pass through the tree, and could only be computed with information available in the trained tree.

 * If we *only* need feature importances for training the XGBoost model and don't care to provide them to the user, one idea is to use a different `NumericSplitType` and `CategoricalSplitType` for XGBoost that captures the necessary information to compute feature importance in the `AuxiliarySplitInfo` class.
   - Of course, this `AuxiliarySplitInfo` is going to be present in every single `DecisionTree` node, so making it hold members will make the XGBoost model larger.
   - It would be possible to "convert" a final XGBoost model where we no longer need the feature importances into a type that does not hold anything in the `AuxiliarySplitInfo`.

 * Adding template parameters to `DecisionTree` is a possibility but something I really really would like to avoid, because:
   - We have to update all downstream documentation.
   - All downstream classes that use `DecisionTree` need to be updated.
   - We would need to think extremely carefully to ensure that whatever we add is flexible enough to support potential future uses.

 * If `XGBoost` is sufficiently different, perhaps it makes sense to reuse the `DecisionTree` class internally, but as a wrapper?  e.g. imagine an `XGBoost` class that holds a `DecisionTree` object.  It's okay now if this `XGBoost` class holds other things, because there is only one `XGBoost` object (whereas for a `DecisionTree` object, there is one `DecisionTree` per node in the tree).

These are not the only three options, but they are rough outlines of the only three that I can see that are feasible.  In short, the primary constraint is that we must implement XGBoost in a way that is not invasive to the `DecisionTree` class, and in situations where we cannot avoid changing `DecisionTree`, we must ensure that we are not breaking the way the `DecisionTree` works or introducing an awkward API that is really XGBoost-specific and doesn't make sense when used with the regular `DecisionTree` class.

I hope this discussion is helpful, please let me know if I can clarify anything, and please make sure all the ideas I wrote above make sense.  I know that there is a lot of context to mlpack; it has been around for over 15 years and is quite large, so it is kind of unavoidable that there is a lot to keep in mind to keep things consistent.

**Comment by TirelessClock** [comment]
@rcurtin I was thinking, if it was just one or two things, we could've done a wrapper or some kind of workaround implementation. 

The thing is, going forward, XGB (and by extension Gradient Boosting too), could use a lot of changes at the Decision Tree level. The feature additions I suggested above are just the main, big stuff, but there are a lot of other smaller feature additions, and maybe it'd be more convenient to implement a more customised XGBTree, DecisionTree specifically for XGB.

It could give me room to do more with, like put XGB specific limits on depth, splits etc. A lot of this is already implemented in Decision Tree, so accessing it can be easier, and I can just include the details of this in the XGB documentation.
